File: src/main/java/moe/sota/decompiler/controllers/TreeController.java
package moe.sota.decompiler.controllers;

import lombok.Getter;
import moe.sota.decompiler.models.ArchiveModel;
import moe.sota.decompiler.models.BaseModel;
import moe.sota.decompiler.views.TreeView;
import org.jetbrains.annotations.NotNull;

import javax.swing.SwingUtilities;
import javax.swing.tree.DefaultMutableTreeNode;
import javax.swing.tree.DefaultTreeModel;
import javax.swing.tree.TreePath;
import java.util.Collections;
import java.util.List;

public class TreeController extends BaseController<TreeView> {

    @Getter
    private static final TreeController INSTANCE = new TreeController();

    private TreeController() {
        super(new TreeView());
    }

    public void setArchive(ArchiveModel archiveModel) {
        DefaultTreeModel treeModel = (DefaultTreeModel) getView().getTree().getModel();
        DefaultMutableTreeNode treeNode = createTreeNode(archiveModel);
        treeModel.setRoot(treeNode);
        treeModel.reload();

        SwingUtilities.invokeLater(() -> {
            var tree = getView().getTree();
            TreePath topPath = new TreePath(treeNode.getPath());

            tree.expandPath(topPath);

            if (tree.getRowCount() > 0) {
                tree.expandRow(0);
                tree.collapseRow(0);
                tree.expandPath(topPath);
            }

            tree.revalidate();
            tree.repaint();
        });
    }

    private @NotNull DefaultMutableTreeNode createTreeNode(BaseModel baseModel) {
        DefaultMutableTreeNode treeNode = new DefaultMutableTreeNode(baseModel);
        List<BaseModel> children = baseModel.getChildren();
        Collections.sort(children);
        for (BaseModel child : children)
            treeNode.add(createTreeNode(child));
        return treeNode;
    }

}

File: jvm/src/main/kotlin/moe/sota/decompiler/jvm/controllers/TreeController.kt
package moe.sota.decompiler.jvm.controllers

import java.util.*
import javax.swing.SwingUtilities
import javax.swing.tree.DefaultMutableTreeNode
import javax.swing.tree.DefaultTreeModel
import javax.swing.tree.TreePath
import moe.sota.decompiler.jvm.models.ArchiveModel
import moe.sota.decompiler.jvm.models.BaseModel
import moe.sota.decompiler.jvm.views.TreeView

class TreeController(private val treeView: TreeView) {
    fun setArchive(archiveModel: ArchiveModel) {
        val treeModel = treeView.tree.model as DefaultTreeModel
        val rootNode = treeModel.getRoot() as DefaultMutableTreeNode
        val treeNode = createTreeNode(archiveModel)
        rootNode.removeAllChildren()
        rootNode.add(treeNode)
        treeModel.reload()

        SwingUtilities.invokeLater { treeView.tree.expandPath(TreePath(treeNode.path)) }
    }

    private fun createTreeNode(baseModel: BaseModel): DefaultMutableTreeNode {
        val treeNode = DefaultMutableTreeNode(baseModel)
        val children = baseModel.children
        Collections.sort(children)
        for (child in children) treeNode.add(createTreeNode(child))
        return treeNode
    }
}
